### DISCUSSION TITLE
Feature request : Auto saves

### DISCUSSION THREAD

**Discussion by Mudloop** [SynthstromAudible/DelugeFirmware#882] [Ideas]
Would be great if the deluge periodically auto-saves a version of your current project somewhere, time stamped or something.

and would be even greater if on boot, it recalls the latest saved state. So you can just shut it off at any time and dive back in without having to deal with saving / loading all the time.

**Comment by darenager** [comment]
Do you mean a separate version is auto saved @Mudloop ?  I think it would be good to separate user saves and auto saves.

**Comment by Mudloop** [comment]
Yes, separate. The other night (after posting this), I lost a track because I tried loading a kit and loaded a project instead. User error of course, but mistakes happen.

Hadn’t saved once. So it would be great if the deluge had taken it upon itself to save a version at some point.

I’m not entirely sure how it should work with audio clips because I can imagine those can fill up the SD quickly. Maybe only keep one or two autosaves of the project, or maybe it could use a hashing system to save audio files, so it doesn’t need to save the same one twice? If CURRENTFILEHASH.wav exists, it can skip saving that. But I’m unfamiliar with the deluge’s inner workings so I’m not sure how feasible that would be. If content hashes have the potential to be too slow, maybe a timestamp of when it was recorded, or just a unique ID would do the trick. Just spitballing tho

**Reply by darenager** [reply]
I believe the audio clips are saved as they are recorded, separate from the project, so should not be an issue.

**Comment by runawaykiwi** [comment]
FYI, I broke out the 'on boot, it recalls the latest saved state' suggestion into a separate discussion since that can be implemented separately from auto saving : https://github.com/SynthstromAudible/DelugeFirmware/discussions/902

**Comment by gramster** [comment]
Time-stamped saves would be nice, but the real missing feature IMO is just snapshotting song state after edits and restoring the state when powering up. Just like the Elektron boxes and Polyend Play+. There's overlap with https://github.com/SynthstromAudible/DelugeFirmware/discussions/902, so maybe this is a dup.

**Comment by o010b** [comment]
Suggestion: For now, autosave at a set interval to a file recovery.xml. It's better than nothing when the community firmware is shaky and can be improved upon later.

**Comment by m-m-adams** [comment]
I'm gonna start working on this now that the task manager is working solidly!
